File: HomelabSwift/Homelab/Services/BackupModels.swift
import Foundation

// MARK: - Backup Envelope

struct BackupEnvelope: Codable {
    let version: Int
    let exportedAt: String
    let appVersion: String
    let services: [BackupServiceEntry]

    static let currentVersion = 1
}

// MARK: - Backup Service Entry

struct BackupServiceEntry: Codable {
    let type: String
    let label: String
    let url: String
    let token: String?
    let username: String?
    let apiKey: String?
    let piholePassword: String?
    let piholeAuthMode: String?
    let fallbackUrl: String?
    let allowSelfSigned: Bool
    let password: String?
    let isPreferred: Bool
}

// MARK: - ServiceType ↔ Backup String Mapping

enum BackupServiceTypeMapper {

    /// Canonical lowercase backup key for a given ServiceType.
    static func backupKey(for type: ServiceType) -> String {
        switch type {
        case .portainer:         return "portainer"
        case .pihole:            return "pihole"
        case .adguardHome:       return "adguard_home"
        case .technitium:        return "technitium"
        case .beszel:            return "beszel"
        case .healthchecks:      return "healthchecks"
        case .linuxUpdate:            return "linux_update"
        case .dockhand:               return "dockhand"
        case .gitea:             return "gitea"
        case .nginxProxyManager: return "nginx_proxy_manager"
        case .pangolin:          return "pangolin"
        case .patchmon:          return "patchmon"
        case .jellystat:         return "jellystat"
        case .plex:              return "plex"
        case .radarr:            return "radarr"
        case .sonarr:            return "sonarr"
        case .lidarr:            return "lidarr"
        case .qbittorrent:       return "qbittorrent"
        case .jellyseerr:        return "jellyseerr"
        case .prowlarr:          return "prowlarr"
        case .bazarr:            return "bazarr"
        case .gluetun:           return "gluetun"
        case .flaresolverr:      return "flaresolverr"
        }
    }

    /// Resolve a backup key string back to a ServiceType, or nil if unknown.
    static func serviceType(from key: String) -> ServiceType? {
        let normalized = key.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        switch normalized {
        case "portainer":            return .portainer
        case "pihole":               return .pihole
        case "adguard_home",
             "adguardhome":          return .adguardHome
        case "technitium",
             "technitium_dns",
             "technitium-dns":       return .technitium
        case "beszel":               return .beszel
        case "healthchecks":         return .healthchecks
        case "linux_update",
             "linuxupdate",
             "linux-update":          return .linuxUpdate
        case "dockhand":              return .dockhand
        case "gitea":                return .gitea
        case "nginx_proxy_manager",
             "nginxproxymanager":     return .nginxProxyManager
        case "pangolin":             return .pangolin
        case "patchmon":             return .patchmon
        case "jellystat":            return .jellystat
        case "plex":                 return .plex
        case "radarr":               return .radarr
        case "sonarr":               return .sonarr
        case "lidarr":               return .lidarr
        case "qbittorrent":          return .qbittorrent
        case "jellyseerr":           return .jellyseerr
        case "prowlarr":             return .prowlarr
        case "bazarr":               return .bazarr
        case "gluetun":              return .gluetun
        case "flaresolverr":         return .flaresolverr
        default:                     return nil
        }
    }

    /// PiHoleAuthMode string mapping.
    static func piholeAuthMode(from string: String?) -> PiHoleAuthMode? {
        guard let string else { return nil }
        switch string.lowercased() {
        case "session": return .session
        case "legacy":  return .legacy
        default:        return nil
        }
    }

    static func backupAuthMode(for mode: PiHoleAuthMode?) -> String? {
        guard let mode else { return nil }
        switch mode {
        case .session: return "session"
        case .legacy:  return "legacy"
        }
    }
}

// MARK: - Conversion Helpers

extension ServiceInstance {
    /// Create a BackupServiceEntry from this instance.
    func toBackupEntry(isPreferred: Bool) -> BackupServiceEntry {
        BackupServiceEntry(
            type: BackupServiceTypeMapper.backupKey(for: type),
            label: displayLabel,
            url: url,
            token: token.isEmpty ? nil : token,
            username: username,
            apiKey: apiKey,
            piholePassword: piholePassword,
            piholeAuthMode: BackupServiceTypeMapper.backupAuthMode(for: piholeAuthMode),
            fallbackUrl: fallbackUrl,
            allowSelfSigned: allowSelfSigned,
            password: password,
            isPreferred: isPreferred
        )
    }
}

extension BackupServiceEntry {
    /// Convert back to a ServiceInstance. Returns nil if the service type is unknown.
    func toServiceInstance() -> ServiceInstance? {
        guard let serviceType = BackupServiceTypeMapper.serviceType(from: type) else { return nil }
        return ServiceInstance(
            id: UUID(),
            type: serviceType,
            label: label,
            url: url,
            token: token ?? "",
            username: username,
            apiKey: apiKey,
            piholePassword: piholePassword,
            piholeAuthMode: BackupServiceTypeMapper.piholeAuthMode(from: piholeAuthMode),
            fallbackUrl: fallbackUrl,
            allowSelfSigned: allowSelfSigned,
            password: password
        )
    }
}

File: HomelabAndroid/app/src/main/java/com/homelab/app/domain/model/BackupModels.kt
package com.homelab.app.domain.model

import com.homelab.app.util.ServiceType
import kotlinx.serialization.Serializable
import java.util.UUID

// MARK: - Backup Envelope

@Serializable
data class BackupEnvelope(
    val version: Int,
    val exportedAt: String,
    val appVersion: String,
    val services: List<BackupServiceEntry>
) {
    companion object {
        const val CURRENT_VERSION = 1
    }
}

// MARK: - Backup Service Entry

@Serializable
data class BackupServiceEntry(
    val type: String,
    val label: String,
    val url: String,
    val token: String? = null,
    val username: String? = null,
    val apiKey: String? = null,
    val piholePassword: String? = null,
    val piholeAuthMode: String? = null,
    val fallbackUrl: String? = null,
    val allowSelfSigned: Boolean,
    val password: String? = null,
    val isPreferred: Boolean
)

// MARK: - ServiceType ↔ Backup String Mapping

object BackupServiceTypeMapper {

    /** Canonical lowercase backup key for a given ServiceType. */
    fun backupKey(type: ServiceType): String {
        return when (type) {
            ServiceType.PORTAINER -> "portainer"
            ServiceType.PIHOLE -> "pihole"
            ServiceType.ADGUARD_HOME -> "adguard_home"
            ServiceType.TECHNITIUM -> "technitium"
            ServiceType.BESZEL -> "beszel"
            ServiceType.HEALTHCHECKS -> "healthchecks"
            ServiceType.LINUX_UPDATE -> "linux_update"
            ServiceType.DOCKHAND -> "dockhand"
            ServiceType.CRAFTY_CONTROLLER -> "crafty_controller"
            ServiceType.GITEA -> "gitea"
            ServiceType.NGINX_PROXY_MANAGER -> "nginx_proxy_manager"
            ServiceType.PANGOLIN -> "pangolin"
            ServiceType.PATCHMON -> "patchmon"
            ServiceType.JELLYSTAT -> "jellystat"
            ServiceType.PLEX -> "plex"
            ServiceType.RADARR -> "radarr"
            ServiceType.SONARR -> "sonarr"
            ServiceType.LIDARR -> "lidarr"
            ServiceType.QBITTORRENT -> "qbittorrent"
            ServiceType.JELLYSEERR -> "jellyseerr"
            ServiceType.PROWLARR -> "prowlarr"
            ServiceType.BAZARR -> "bazarr"
            ServiceType.GLUETUN -> "gluetun"
            ServiceType.FLARESOLVERR -> "flaresolverr"
            ServiceType.WAKAPI -> "wakapi"
            ServiceType.UNKNOWN -> "unknown"
        }
    }

    /** Resolve a backup key string back to a ServiceType, or null if unknown. */
    fun serviceType(key: String): ServiceType? {
        val normalized = key.lowercase().trim()
        return when (normalized) {
            "portainer" -> ServiceType.PORTAINER
            "pihole" -> ServiceType.PIHOLE
            "adguard_home", "adguardhome" -> ServiceType.ADGUARD_HOME
            "technitium", "technitium_dns", "technitium-dns" -> ServiceType.TECHNITIUM
            "beszel" -> ServiceType.BESZEL
            "healthchecks" -> ServiceType.HEALTHCHECKS
            "linux_update", "linuxupdate", "linux-update" -> ServiceType.LINUX_UPDATE
            "dockhand" -> ServiceType.DOCKHAND
            "gitea" -> ServiceType.GITEA
            "nginx_proxy_manager", "nginxproxymanager" -> ServiceType.NGINX_PROXY_MANAGER
            "pangolin" -> ServiceType.PANGOLIN
            "patchmon" -> ServiceType.PATCHMON
            "jellystat" -> ServiceType.JELLYSTAT
            "plex" -> ServiceType.PLEX
            "radarr" -> ServiceType.RADARR
            "sonarr" -> ServiceType.SONARR
            "lidarr" -> ServiceType.LIDARR
            "qbittorrent" -> ServiceType.QBITTORRENT
            "jellyseerr" -> ServiceType.JELLYSEERR
            "prowlarr" -> ServiceType.PROWLARR
            "bazarr" -> ServiceType.BAZARR
            "gluetun" -> ServiceType.GLUETUN
            "flaresolverr" -> ServiceType.FLARESOLVERR
            "wakapi" -> ServiceType.WAKAPI
            "crafty_controller", "crafty" -> ServiceType.CRAFTY_CONTROLLER
            else -> null
        }
    }

    /** PiHoleAuthMode string mapping. */
    fun piholeAuthMode(string: String?): PiHoleAuthMode? {
        return when (string?.lowercase()) {
            "session" -> PiHoleAuthMode.SESSION
            "legacy" -> PiHoleAuthMode.LEGACY
            else -> null
        }
    }

    fun backupAuthMode(mode: PiHoleAuthMode?): String? {
        return when (mode) {
            PiHoleAuthMode.SESSION -> "session"
            PiHoleAuthMode.LEGACY -> "legacy"
            null -> null
        }
    }
}

// MARK: - Conversion Helpers

fun ServiceInstance.toBackupEntry(isPreferred: Boolean): BackupServiceEntry {
    return BackupServiceEntry(
        type = BackupServiceTypeMapper.backupKey(type),
        label = label, // in iOS it was displayLabel, here label is equivalent
        url = url,
        token = token.ifBlank { null },
        username = username,
        apiKey = apiKey,
        piholePassword = piholePassword,
        piholeAuthMode = BackupServiceTypeMapper.backupAuthMode(piholeAuthMode),
        fallbackUrl = fallbackUrl,
        allowSelfSigned = allowSelfSigned,
        password = password,
        isPreferred = isPreferred
    )
}

fun BackupServiceEntry.toServiceInstance(): ServiceInstance? {
    val mappedType = BackupServiceTypeMapper.serviceType(type) ?: return null
    return ServiceInstance(
        id = UUID.randomUUID().toString(),
        type = mappedType,
        label = label,
        url = url,
        token = token ?: "",
        username = username,
        apiKey = apiKey,
        piholePassword = piholePassword,
        piholeAuthMode = BackupServiceTypeMapper.piholeAuthMode(piholeAuthMode),
        fallbackUrl = fallbackUrl,
        allowSelfSigned = allowSelfSigned,
        password = password
    )
}
